File: services/webhook-service/webhook_service/presentation/http/envelope.go
package http

import (
	"encoding/json"
	"log/slog"
	"net/http"
	"time"
)

// The response envelope every service in this API shares: a success carries
// `data` + `meta`, a failure carries `error` + `meta`.
type envelope struct {
	Data  any            `json:"data,omitempty"`
	Error *errorBody     `json:"error,omitempty"`
	Meta  map[string]any `json:"meta"`
}

type errorBody struct {
	Code    string        `json:"code"`
	Message string        `json:"message"`
	Details []errorDetail `json:"details,omitempty"`
}

type errorDetail struct {
	Field   string `json:"field"`
	Code    string `json:"code"`
	Message string `json:"message"`
}

const correlationHeader = "X-Correlation-ID"

func writeOK(writer http.ResponseWriter, data any, meta map[string]any) {
	if meta == nil {
		meta = map[string]any{}
	}

	writeJSON(writer, http.StatusOK, envelope{Data: data, Meta: meta})
}

func writeError(
	writer http.ResponseWriter,
	request *http.Request,
	status int,
	code string,
	message string,
	details ...errorDetail,
) {
	writeJSON(writer, status, envelope{
		Error: &errorBody{Code: code, Message: message, Details: details},
		Meta: map[string]any{
			"request_id": request.Header.Get(correlationHeader),
			"timestamp":  time.Now().UTC().Format(isoLayout),
		},
	})
}

func writeJSON(writer http.ResponseWriter, status int, body envelope) {
	writer.Header().Set("Content-Type", "application/json")
	writer.WriteHeader(status)

	if encodeErr := json.NewEncoder(writer).Encode(body); encodeErr != nil {
		slog.Error("failed to encode response", "error", encodeErr)
	}
}

File: services/write-service/write_service/common/http_contract/envelope.py
from typing import Any

from correlation import get_correlation_id
from django.utils import timezone
from rest_framework.response import Response

from .codes import STATUS_FOR_ERROR_CODE, ErrorCode
from .exceptions import ErrorDetail


def error_meta() -> dict[str, Any]:
    return {
        "request_id": get_correlation_id(),
        "timestamp": timezone.now().isoformat(),
    }


def ok_payload(data: Any, meta: dict[str, Any] | None = None) -> dict[str, Any]:
    return {"data": data, "meta": meta if meta is not None else {}}


def error_payload(
    code: ErrorCode,
    message: str,
    details: list[ErrorDetail] | None = None,
) -> dict[str, Any]:
    error: dict[str, Any] = {"code": str(code), "message": message}

    if details and code is not ErrorCode.INTERNAL_ERROR:
        error["details"] = [detail.as_dict() for detail in details]

    return {"error": error, "meta": error_meta()}


def ok(
    data: Any,
    meta: dict[str, Any] | None = None,
    *,
    status_code: int = 200,
    headers: dict[str, str] | None = None,
) -> Response:
    return Response(ok_payload(data, meta), status=status_code, headers=headers)


def fail(
    code: ErrorCode,
    message: str,
    *,
    details: list[ErrorDetail] | None = None,
    status_code: int | None = None,
    headers: dict[str, str] | None = None,
) -> Response:
    return Response(
        error_payload(code, message, details),
        status=status_code or STATUS_FOR_ERROR_CODE[code],
        headers=headers,
    )
